From 91eaa9221b45b506d7a5d84a9ec8c204b5c8c34f Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Fri, 8 May 2026 13:13:37 +0800
Subject: [PATCH] calendar-server: Fix build failure with libical 4

Fix the following build error with libical 4 due to function signature changes:

    error: passing argument 4 of 'get_time_from_property' from incompatible pointer type [-Wincompatible-pointer-types]
    note: expected 'ICalTime * (*)(ICalProperty *)' but argument is of type 'ICalTime * (*)(const ICalProperty *)'

Link: https://github.com/libical/libical/blob/4.0/docs/MigrationGuide_to_4.md
Part-of: <https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/4215>

Only the gnome-shell-calendar-server.c hunk of the upstream commit is taken;
src/calendar-server/reminder-watcher.c does not exist in the 50.x series.

Upstream-Status: Backport [https://gitlab.gnome.org/GNOME/gnome-shell/-/commit/b34766f3992c7616b7816a47249fb463c053e5fa]
---
 src/calendar-server/gnome-shell-calendar-server.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/calendar-server/gnome-shell-calendar-server.c b/src/calendar-server/gnome-shell-calendar-server.c
index 0b2f88eb7..622c16181 100644
--- a/src/calendar-server/gnome-shell-calendar-server.c
+++ b/src/calendar-server/gnome-shell-calendar-server.c
@@ -112,11 +112,17 @@ typedef struct
   time_t  end_time;
 } CalendarAppointment;
 
+#if ICAL_CHECK_VERSION(4, 0, 0)
+typedef ICalTime * (* GetICalPropFunc) (const ICalProperty *prop);
+#else
+typedef ICalTime * (* GetICalPropFunc) (ICalProperty *prop);
+#endif
+
 static gboolean
 get_time_from_property (ECalClient            *cal,
                         ICalComponent         *icomp,
                         ICalPropertyKind       prop_kind,
-                        ICalTime * (* get_prop_func) (ICalProperty *prop),
+                        GetICalPropFunc        get_prop_func,
                         ICalTimezone          *default_zone,
                         ICalTime              **out_itt,
                         ICalTimezone          **out_timezone)
